### PR TITLE
Make preprocessing and postprocessing consistent accross transforms

### PR DESCRIPTION
**Pull request by felixblanke**
This addresses #92.

For all discrete transforms, the preprocessing and postprocessing of coefficients and tensors is very similar (i.e. folding and swapping of axes, adding batch dims, etc.). This PR moves this functionality into shared functions that use `_map_result`.

Also the check for consistent devices and dtypes between coefficient tensors is moved into a function in `_utils`.

Last, as it was possible to add it with a few lines of code, I added the $n$-dimensional fully separable transforms (`fswavedecn`, `fswaverecn`). If this is not wanted, I can revert their addition.

Further, I did some minor refactorings along the way.

### CODE REVIEW

**Comment by v0lta** [comment]
I did not add n-dimensional separable transforms on propose, because I was thinking people will ask for these in all other cases, too, where these are trickier to deliver.

File: src/ptwt/_util.py [hunk 1]
```diff
@@ -186,6 +187,32 @@ def _check_axes_argument(axes: Sequence[int]) -> None:
...
+    return tensor
+
+
+def _check_same_device_dtype(
+    coeffs: Union[list[torch.Tensor], WaveletCoeff2d, WaveletCoeffNd],
+) -> tuple[torch.device, torch.dtype]:
+    c = _check_if_tensor(coeffs[0])
+    torch_device, torch_dtype = c.device, c.dtype
+
+    _map_result(coeffs, partial(_check_same_device, torch_device=torch_device))
```
Comment (cthoyt) [inline src/ptwt/_util.py:210]: this code would be much easier to understand if you just wrote out an explicit loop

File: src/ptwt/_util.py [hunk 2]
```diff
@@ -245,11 +280,225 @@ def _map_result(
...
+    elif len(ds) > ndim + 1:
+        unfold_axes_fn = partial(_unfold_axes, ds=ds, keep_no=ndim)
+        coeffs = _map_result(coeffs, unfold_axes_fn)
+
+    if tuple(axes) != tuple(range(-ndim, 0)):
+        if len(axes) != ndim:
+            raise ValueError(f"{ndim}D transforms work with {ndim} axes.")
+        else:
+            undo_swap_fn = partial(_undo_swap_axes, axes=axes)
+            coeffs = _map_result(coeffs, undo_swap_fn)
```
Comment (cthoyt) [inline src/ptwt/_util.py:467]: I would really advise against all of these `_map_result` calls - have one function that does the processing that can be reused, then just do list comprehensions for all successive function calls.

```
coeffs = _map_result(coeffs, lambda x: x.squeeze(0))
```

will always be less readable and understandable than

```
coeffs = _map_result(coeffs)
coeffs = [coeff.squeeze(0) for coeff in coeffs]
```

when _map_result has lots of hidden functionality

File: src/ptwt/_util.py (hunk 1 above)
Comment (felixblanke) [reply src/ptwt/_util.py:210]: Fair point :+1:

File: src/ptwt/_util.py (hunk 2 above)
Comment (felixblanke) [reply src/ptwt/_util.py:467]: The snippet
```
coeffs = _map_result(coeffs, lambda x: x.squeeze(0))
```
applies the function `x.squeeze(0)` to all tensors in `coeffs`. In the 1d case (where `coeffs` is of type `list[Tensor]`) this is equivalent to the list comprehension, as you said. However, `coeffs` might also be
* `(Tensor, dict[str, Tensor], ...)`
* `(Tensor, (Tensor, Tensor, Tensor), ...)`

So using `_map_result` allows to write the function once for all possible coefficient types. Would it perhaps help to rename `_map_result` or add documentation for it?

File: src/ptwt/_util.py (hunk 1 above)
Comment (felixblanke) [reply src/ptwt/_util.py:210]: The snippet runs the check on all tensors in the coeff data structure.
Similar as below, adding a simple loop would effectively require to copy the `_map_result` functionality.
We could duplicate this function once for each coeff data type to make it more explicit. I would however prefer to address this by making this call more understandable

**Comment by felixblanke** [comment]
@v0lta I made the n-dim transform private. Does that work?

**Comment by v0lta** [comment]
Yes, that works. However, we won't be able to support n-dimensional transforms across the board because PyTorch does not provide the interfaces we would need to do that. Padding, for example, works only up until 3D ( https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html ). We have the same problem with isotropic convolution. So, I think we should communicate that nd-transforms are out of the scope of this project.

File: src/ptwt/_util.py (hunk 1 above)
Comment (v0lta) [reply src/ptwt/_util.py]: I was a big fan of the _map_result function. We are missing context here. The idea comes from https://jax.readthedocs.io/en/latest/_autosummary/jax.tree.map.html#jax.tree.map, which is a very powerful and reusable concept. We are dealing with a tree-like data structure and want to process all its leaves. I think a tree-map is the right call.

File: src/ptwt/_util.py (hunk 2 above)
Comment (v0lta) [reply src/ptwt/_util.py]: I think the name should have something to do with tree and map. The new `_apply_to_tensor_elems` name really hides how general the concept is. I would take a page from https://jax.readthedocs.io/en/latest/_autosummary/jax.tree.map.html#jax.tree.map and also use their type hinting. The concept does not exist in torch, but I think it makes sense here, since we save on a lot of boilerplate-code. Perhaps we should include the link and explain what's going on?

**Comment by v0lta** [comment]
In general, I am a big fan of this full request! Thanks at @felixblanke  I am going to clean up the docs for map_result and commit here.

File: src/ptwt/_util.py (hunk 1 above)
Comment (v0lta) [reply src/ptwt/_util.py]: We inherit Lists of dicts or tuples as coefficient data type from `pywt`.  Having some kind of a tree map function helps us since it will allow us to process the leaves in every case without having to loop through all of these different data types all of the time. I don't think we want the jax dependency just for this one function. So I think it's ok to have our own, especially since it's not user-facing.

File: src/ptwt/_util.py (hunk 2 above)
Comment (v0lta) [reply src/ptwt/_util.py]: Here is an interesting intro discussing the pytree processing philosophy: https://jax.readthedocs.io/en/latest/working-with-pytrees.html .

**Comment by v0lta** [comment]
Our `coeff_tree_map` is not a general tree map, but it does not have to be since we know the approximation tensor will always be the first entry. I ran the `not-slow` tests; everything checked out. The code is cleaner now. If everyone is on board, I would be ready to merge.

**Comment by felixblanke** [comment]
I think we so far only refer to the Packet data structure as a tree. Maybe we can add a link to the JAX discussion as a reference?

**Comment by v0lta** [comment]
I am not sure if users need to know. I think this is more for us here internally. Unlike Jax's tree map, ours is coefficient-specific, hence the proposed `coeff_tree_map` function name.

**Comment by v0lta** [comment]
Actually, never mind, the user argument does not matter since it's a private function. If you think it helps to understand the idea please add a link. I think it might help potential future contributors.

File: src/ptwt/_util.py (hunk 2 above)
Comment (v0lta) [reply src/ptwt/_util.py]: I also think @cthoyt has a point since the tree-map concept is not very popular.

Comment (v0lta) [reply src/ptwt/_util.py]: I agree when comparing

``` python
coeffs = _map_result(coeffs, partial(torch.squeeze, 0))
````
to 
``` python
coeffs = [coeff.squeeze(0) for coeff in coeffs]
```
The list wins, but what if it's a nested structure?

**Comment by v0lta** [comment]
Okay, I think we are ready to merge.

File: src/ptwt/separable_conv_transform.py
Comment (felixblanke) [inline src/ptwt/separable_conv_transform.py:1]: I think we should probably move the Note down to `_fswavedecn`

Comment (v0lta) [reply src/ptwt/separable_conv_transform.py:1]: please do!

Comment (v0lta) [reply src/ptwt/separable_conv_transform.py:1]: done

**Comment by v0lta** [comment]
okay let's merge!
